c7n/tags.py

```diff
@@ -1020,8 +1020,11 @@ def validate(self):
         return self
 
     def process(self, resources):
-        related_resources = dict(
```

Comment:
isn't the simpler fix to just remove the dict and leave the zipped results here? that avoids a nested loop in the processing below.

Reply:
alternatively another simpler impl is just reverse the k:v in the dict to be v:k

Reply:
i was looking at that and that would work if every resource had a value at the user defined key location but in some cases it doesn't which leads to a mismatch of related resource and original resource.

In this case, eni's may be unattached and thus missing the `Attachment.InstanceId` key in the resource json, which can result in the following:

```python
>>> related_resources = list(zip(jmespath.search('[].%s' % self.data['key'], resources), resources))
>>> print([r[1].get('Attachment', 'Missing') for r in related_resources)
['Missing', 'Missing']
```

Reply:
the jmespath search would return a None in that case, so the zip is still balanced on cardinality. also in that case it would need to be skipped in the loop, ie that case has to be handled anyways, which i'm not seeing in the pr?

Reply:
ah.. i see what you mean..

```python
>>> x = [{'k': 1}, {'k': 2}, {'a': 1}, {'k': 3}]
>>> list(zip(jmespath.search('[].k', x), x))
[(1, {'k': 1}), (2, {'k': 2}), (3, {'a': 1})]
```

Reply:
i've been exploring on jmespath syntax to see if we can preserve the null values better, here's one example i found
```python
>>> x
[{'k': 1}, {'k': 2}, {'a': 1}, {'k': 3}]
>>> jmespath.search('[].[k]', x)
[[1], [2], [None], [3]]
```

Reply:
one more expression, there is special handling it in jmespath around null, so besides the nesting in a list to preserve null, for a flat structure the other option is substituting in a known missing value like (can be integer, string, bool, etc)

```python
jmespath.search('[].[k || `false`]|[]', x)
[1, 2, False, 3]

```